Implementation for insert method for SQLAlchemy backends

harsharaj96 opened the pull request:
Closes #2613

The implementation is similar to what is currently implemented for the Impala database. Please have a look and merge it if you think this helps your requirement. I have also added commentary to what will work and what won't in the code so that enhancements can be made to it later on. Thanks.

pep8speaks commented:
Hello @harsharaj96! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`ibis/backends/sqlite/client.py`](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py):

> [Line 418:71](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L418): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 419:9](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L419): [E125](https://duckduckgo.com/?q=pep8%20E125) continuation line with same indent as next logical line
> [Line 419:52](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L419): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 419:53](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L419): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 429:80](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L429): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 430:80](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L430): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)
> [Line 431:80](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L431): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 431:85](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L431): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 438:80](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L438): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 438:93](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L438): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 439:80](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L439): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 439:91](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L439): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 448:80](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L448): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 79 characters)
> [Line 455:80](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L455): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 79 characters)
> [Line 458:80](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L458): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 465:80](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L465): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 465:87](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L465): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 466:17](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L466): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 469:80](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L469): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 469:87](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L469): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 470:17](https://github.com/ibis-project/ibis/blob/86ef1c38778c659f80ca2616b997381ff6b95b63/ibis/backends/sqlite/client.py#L470): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

datapythonista commented:
Thanks for the contribution @harsharaj96. I'll have a closer look later, but to give a bit more context. SQLite, PosgreSQL, and MySQL are the backends that use SQLAlchemy internally. Didn't check if PostgreSQL and MySEL have the insert method, but I think it should be possible to have a single insert for the 3 of them.

Do you mind doing some more researh on it?

Also, for this PR we'll need:
- Add a test for the new `insert` method
- A a line in the release notes for this change
- Optionally, in the tutorials we've got a part where we comment that inserting is possible in Ibis, but we don't have an example because it doesn't work in Ibis. Adding an example there would be nice too (we can surely do it in a follow up PR, don't worry if you don't feel like doing it here).

datapythonista commented:
Also, before we can merge, you'll have to run `black .` and `isort .`, which will make some automatic changes to the files, and commit those changes, so the CI is green.

harsharaj96 commented:
Hi @datapythonista,
Firstly, thanks for your response. Based on lookup, PostgresSQL and MySQL clients in this repo don't have `insert` methods implemented. Having a common `insert` method implemented in the SQLAlchemy sounds good but I am confused about how it is different from the `load_data` method that is already implemented which uses a dataframe to load data into a table.

harsharaj96 commented:
I have run the `black .` and `isort .` commands and made changes to the file. Seems like the CI checks also passed now. Thanks for the inputs.

harsharaj96 commented:
Regarding the test case for `insert` method. I haven't totally understood the other test cases for SQLite, like table creation and loading data. I'll try to add a test case once I understand them. If you can point me to any reference example usages or starts points, it would be great. Thanks in advance @datapythonista.

datapythonista commented:
You can check the clickhouse backend as reference. In the `client.py` and `test_client.py` you have an implementation and a test.

`load_data` translates to SQL `LOAD DATA` while `insert` translate to SQL `INSERT`. Ibis just wraps that, and let users choose which SQL statement is more convenient.

Besides the test, you are also missing here an entry in `docs/source/release/index.rst`.

Thanks!

datapythonista commented:
@harsharaj96 do you have time to add the test and the entry in the release notes?

harsharaj96 commented:
> @harsharaj96 do you have time to add the test and the entry in the release notes?

Hi @datapythonista ,
I'm so sorry, for some reason I missed your previous msg/mail. I'll try to work on it today and let you know. 
Thanks.

harsharaj96 commented:
Hi @datapythonista,

I have now added test cases for insert (two functions with allowed functionalities, insert using a dataframe and insert using a table). I have also added the entry in the release note (in the sqlite_insert PR branch itself). Please take a look and if you think everything is fine then please merge this PR.

Thanks!

datapythonista commented on ibis/backends/sqlite/client.py:
```diff
@@ -412,3 +413,88 @@ def _table_from_schema(
         columns = self._columns_from_schema(name, schema)
         return sa.Table(name, self.meta, schema=database, *columns)
+
+    def insert(
```
Did you check if this could be implemented in the base SQLAlchemy client, so it also works for Postgres and MySQL?

harsharaj96 replied:
I think it can be implemented. Should I do it in this PR itself? Also, this might require updating of test cases as well.

jreback replied:
yes pls do

harsharaj96 replied:
@datapythonista, @jreback please take a look now. Have added (moved and modified) this function to both the places relevant under the `AlchemyClient` class under the `backends` directory.

datapythonista reviewed (approved):
Looks good, you'll have to merge master to resolve the conflicts.

harsharaj96 commented on docs/source/release/index.rst:
```diff
@@ -12,6 +12,7 @@ Release Notes
    These release notes are for versions of ibis **1.0 and later**. Release
    notes for pre-1.0 versions of ibis can be found at :doc:`release-pre-1.0`
 
+* :feature:`2613` Implement .insert() for SQLite
```
Done!

datapythonista commented on ibis/backends/base_sqlalchemy/alchemy.py:
```diff
@@ -0,0 +1,1741 @@
+import contextlib
```
I think this has been committed by mistake, this file was deleted.

datapythonista commented on ibis/backends/sqlite/tests/test_client.py:
```diff
@@ -116,3 +116,55 @@ def test_verbose_log_queries(con):
     assert query == expected
+
+
+def test_insert_from_dataframe(con, df):
```
This should be tested for all backends. Maybe you just need to find an `insert` test in `ibis/backends/tests` where the test is skipped for the Alchemy backends.

harsharaj96 replied:
I couldn't find any `insert` function in `ibis/backends/tests` directory which is being skipped. But I did find a few tests related to `load_data` and `create_table` etc but couldn't understand where it was being orchestrated (called) from. 
If you either find the `insert` test or if you still want me to assume the test function input (arguments) and write a test for `insert` then let me know. @datapythonista 

Thanks!

datapythonista replied:
This should be the file to test it: `ibis/backends/tests/test_client.py`

And you can use `@pytest.mark.only_on_backends(['sqlite', 'postgres', 'mysel'])`. And I think Impala and maybe other support this, so you can add them to the list too.

harsharaj96 replied:
Added the test to this file with `@pytest.mark.only_on_backends(['sqlite', 'postgres', 'mysql'], reason="run only if backend is SQLAlchemy based",)` since the backends based on SQLAlchemy have `insert` method signature different from others like `impala`.

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,82 @@ def _sqla_table_to_expr(self, table):
+          (table name to insert data from)
+
+        NotImplementedError
+          Inserting with values is not implemented for SQLite
```
Not SQLite anymore.

datapythonista commented on docs/source/release/index.rst:
```diff
@@ -12,6 +12,8 @@ Release Notes
    notes for pre-1.0 versions of ibis can be found at :doc:`release-pre-1.0`
 
+
+* :feature:`2613` Implement `.insert()` for SQLite
```
Can you update, not SQLite only

jreback commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,82 @@ def _sqla_table_to_expr(self, table):
+        database : string, optional
+          name of the attached database that the table is located in.
+        data_obj : pd.Dataframe (Later can also be used for sending in
+        table_expr)
```
formatting?

jreback commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,82 @@ def _sqla_table_to_expr(self, table):
+        from_table_name: Optional[str] = None,
+        if_exists: Optional[str] = 'append',
+        values=None,
+    ):
```
return annotation?

jreback commented:
pls also rebase on master & remove the extra file.

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,82 @@ def _sqla_table_to_expr(self, table):
+                **params,
+            )
+        elif data_obj is None and from_table_name is not None:
+            data_obj = pd.read_sql_table(from_table_name, self.con, **params)
```
This doesn't seem correct. If I call
```python
con.insert(to_table_name='foo', from_table_name='bar')
```
You are getting the data from the database to the client, load it into pandas, and then create an insert with the values returned?

Why not simply run:
```sql
INSERT INTO foo SELECT * FROM bar;
```

If the returned data is large, the time difference among them will be huge.

harsharaj96 replied:
Done!

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,82 @@ def _sqla_table_to_expr(self, table):
+            from_table_name (table name to insert data from)
+
+        NotImplementedError
+            Inserting with values is not implemented for SQLite
```
I think I already added a comment for it. But this is not SQLite specific anymore, will work in any SQLAlchemy backend, so better not use `SQLite`  in the docs, this is misleading. You can use `[...] not implemented for SQLAlchemy based backends` for example.

harsharaj96 replied:
Done!

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,82 @@ def _sqla_table_to_expr(self, table):
+
+        if values is not None:
+            raise NotImplementedError(
+                'Inserting with values is not implemented for SQLite'
```
Same about using SQLite. Do you mind reviewing that it's not mentioned anywhere else?

datapythonista reviewed (commented):
Looking better, added some comments.

datapythonista reviewed (commented):
Thanks @harsharaj96, getting closer. Few more comments, but starts to look good.

harsharaj96 commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,81 @@ def _sqla_table_to_expr(self, table):
+            inserted to to_table_name
+            (Later when table_expr related implementation is done this
+            variable can be used to take in that data)
+        from_table_name: string, optional
```
Done!

harsharaj96 commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,81 @@ def _sqla_table_to_expr(self, table):
+            variable can be used to take in that data)
+        from_table_name: string, optional
+            name of the table from which data needs to be inserted
+        if_exists : string, optional, default 'append'
```
Done! Haven't removed Optional from type hints of the parameter though.

harsharaj96 commented on ibis/backends/sqlite/tests/test_client.py:
```diff
@@ -116,3 +116,48 @@ def test_verbose_log_queries(con):
     assert query == expected
+
+
+def test_insert_from_dataframe(con, df):
```
I thought this test can still stay for the sqlite backend but we can remove it if it's not required.

harsharaj96 commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,81 @@ def _sqla_table_to_expr(self, table):
+            name of the table from which data needs to be inserted
+        if_exists : string, optional, default 'append'
+            The values available are: {‘fail’, ‘replace’, ‘append’}
+        values: None, optional
```
I tried to add it as another way the user can insert data (single row) into a table similar to the `insert` method implemented in impala (`backends/impala/client.py`). This was the idea behind having a parameter for it. If you still feel it's redundant I can remove it.

datapythonista replied:
I see. But yes please, I don't think the Impala `values` is very useful, and makes things too much complex. And in any case, since we're not implementing it, let's better remove the argument.

harsharaj96 replied:
Sure!

harsharaj96 commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,81 @@ def _sqla_table_to_expr(self, table):
+                if_exists=if_exists,
+                **params,
+            )
+        elif data_obj is None and from_table_name is not None:
```
Done!

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,93 @@ def _sqla_table_to_expr(self, table):
+            database = self.database_name
+
+        if isinstance(data_obj, pd.DataFrame) or isinstance(
+            from_table_name, str):
```
This if is not needed. If you add the `else` to the below if/elif should do the same.

harsharaj96 replied:
Sure!

datapythonista commented on ibis/backends/tests/test_client.py:
```diff
@@ -202,3 +203,56 @@ def test_separate_database(con, alternate_current_database, current_data_db):
+    assert len(temporary.execute()) == 0
+    con.insert('temporary_alltypes', data_obj=records)
+
+    temporary_post_insert = con.table('temporary_alltypes')
```
I don't think this should be needed, executing again on `temporary` should return the results.

harsharaj96 replied:
Sure!

datapythonista commented on ibis/backends/tests/test_client.py:
```diff
@@ -202,3 +203,56 @@ def test_separate_database(con, alternate_current_database, current_data_db):
+    ['sqlite', 'postgres', 'mysql'],
+    reason="run only if backend is SQLAlchemy based",
+)
+def test_insert_from_table(con, from_table_name):
```
Do we have a fixture `from_table_name`? Surprising if that's the case.

harsharaj96 replied:
I couldn't find the orchestrator for these testing functions but I was assuming the user will be providing it (`from_table_name` parameter) so that we can insert data from that table.

datapythonista replied:
Those arguments are a pytest magic named fixtures. You define a function `con` that generates one or multiple connections, then you register it as a fixture, and pytest executes the test with the return of the function (it possibly executes it multiple times).

In this case there is no user for the test, is all automated. What I would do is to create two rows manually, and then you create a dataframe for the first test, and you can insert them to a temporary table with `raw_sql` in the second.

harsharaj96 replied:
Updated the tests, please take a look.

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,93 @@ def _sqla_table_to_expr(self, table):
+                    **params,
+                )
+            elif isinstance(from_table_name, str):
+                self.raw_sql(
```
Converting this to SQLAlchemy is pending, I guess you know, but just in case.

harsharaj96 replied:
Working on it.

harsharaj96 commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,81 @@ def _sqla_table_to_expr(self, table):
+                **params,
+            )
+        elif data_obj is None and from_table_name is not None:
+            self.raw_sql(
```
Updated.

datapythonista commented:
Checked quickly and changes look great now. Let me know when this is ready for a last review.

harsharaj96 commented:
Hi @datapythonista,
All the suggested changes are worked on. You can do the review and inform me if I'm missing anything or if something more is required.
Thanks!

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,105 @@ def _sqla_table_to_expr(self, table):
+        database : string, optional
+            name of the attached database that the table is located in.
+        data_obj : pd.Dataframe (Later can also be used for sending in
+        table_expr)
```
I think the indentation here is wrong

harsharaj96 replied:
Wanted to add a comment on the usage of this option in the future. Will add it to the docstring.

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,105 @@ def _sqla_table_to_expr(self, table):
+
+        Raises
+        -------
+        ValueError
```
I think a single Value Error explaining all the different reasons would be better.

harsharaj96 replied:
Okay!

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,105 @@ def _sqla_table_to_expr(self, table):
+            from_table_expr = self.table(from_table_name)
+
+            with self.begin() as bind:
+                to_table.create(bind=bind)
```
Feels like we should be using `if_exists` here, and create, replace or fail depending on its value, right? Would be good to have tests for those too if not too complex.

harsharaj96 replied:
Firstly, `if_exists` must be used on `to_table` and not `from_table`.  
Coming to the usage part, I haven't used `if_exists` here because of a couple of reasons. One is, there's no parameter for it to be included in the `.table()` function so the backend is assuming the table will exist with the given name. Another reason is, I couldn't find functionality in the `.load_data()` and `.create_table()` functions which use another table (`from_table`) for its operations, and hence couldn't implement it similar to that. (This is an additional feature apart from using a dataframe for inserting). Earlier I was converting the `from_table` to a dataframe and then using `to_sql` for insertion which could accept the `if_exists` param and hence it looked fine.  
If you have any thoughts or sample reference which I can follow for the usage of the `if_exists` parameter by using the AlchemyClient operations I'll be happy to update this implementation.
Thanks!

datapythonista replied:
You simply need to do this manually. Something like:
```python
if if_exists == 'fail' and self.exists_table(to_table_name):
    raise RuntimeError('The table already exists')
elif if_exists == 'replace' and self.exists_table(to_table_name):
    self.drop(table_name_from)
...
```

harsharaj96 replied:
Updated.

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,105 @@ def _sqla_table_to_expr(self, table):
+                        )
+                    )
+        else:
+            raise ValueError(
```
Not sure if this should be a TypeError instead, since it's raising mainly when the types don't match.

harsharaj96 replied:
I agree. Must be TypeError.

datapythonista commented on ibis/backends/tests/test_client.py:
```diff
@@ -202,3 +202,97 @@ def test_separate_database(con, alternate_current_database, current_data_db):
+    con.create_table(temp_table, schema=sch)
+    con.load_data(temp_table, df, if_exists='replace')
+    temporary = con.table(temp_table)
+    assert len(temporary.execute()) == 3
```
That's really nice. If I'm not wrong, this setup part is the same in the other test. If it is, instead of reusing you can write a pytest fixture (or a test is you need params or makes more sense) that does this and returns the table name (I think that's what it's needed in the test). What do you think?

harsharaj96 replied:
Now, after updating the test case there's no data to be added to the `to_table` so this won't be relevant. But the idea was a good one.

datapythonista commented on ibis/backends/tests/test_client.py:
```diff
@@ -202,3 +202,97 @@ def test_separate_database(con, alternate_current_database, current_data_db):
+    )
+
+    con.insert('temp_to_table', data_obj=records)
+    assert len(temporary.execute()) == 6
```
I think in your first implementation you compared the dataframes, not only the number of rows in the table. That would be better if possible, so we see that INSERT is inserting the right data in the right columns.

harsharaj96 replied:
Right, that's better. But that requires no data to be present in the `to_table` and only then comparison can be done. Will update the code that way.

datapythonista reviewed (commented):
Some final comments, but looks mostly ready.

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,104 @@ def _sqla_table_to_expr(self, table):
+            data_obj is the dataframe containing data which needs to be
+            inserted to to_table_name
+            (Later when table_expr related implementation is done this
+            variable can be used to take in that data)
```
I guess `table_expr` is an argument you are considering adding in the future. I don't think this is useful for users to know, can you instead create a new issue for that task please?

harsharaj96 replied:
Sure!  
Added #2778

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,104 @@ def _sqla_table_to_expr(self, table):
+            variable can be used to take in that data)
+        from_table_name : string, optional
+            name of the table from which data needs to be inserted
+        if_exists : string, default 'append'
+            The values available are: {‘fail’, ‘replace’, ‘append’}
```
We should explain what this parameter does, something like:

```suggestion
        if_exists : {‘append’, ‘fail’, ‘replace’}
            With append, the rows will be added to the table. With fail, and exception will be raised if the table exists. With replace, if the table exists, it'll be dropped, and recreated before inserting the rows.
```

harsharaj96 replied:
Done!

harsharaj96 commented:
Hi @datapythonista,
All changes are fixed/updated. Please confirm and let me know.
Thanks a lot!

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,138 @@ def _sqla_table_to_expr(self, table):
+                        )
+            elif if_exists == 'append' and to_table_name in self.list_tables():
+                to_table_expr = self.table(to_table_name)
+                to_table_schema = to_table_expr.schema()
```
You are repeating this and other code from the previous `if`. Repeating code is quite a poor development practice. You need to write this code twice, I need to review twice, any change that I suggest you'll have to implement it twice. And future maintenance will also be complicated, inconsistencies, repeating changes...

This is as easy to prevent as just placing the `if` correctly. Or in other cases, to write a function. Can you please do that?

harsharaj96 replied:
Agreed. Structured the conditions in a better way.

datapythonista commented on ibis/backends/tests/test_client.py:
```diff
@@ -202,3 +203,79 @@ def test_separate_database(con, alternate_current_database, current_data_db):
+    con.drop_table(temp_table, force=True)
+    con.create_table(temp_table, schema=sch)
+    temporary = con.table(temp_table)
+    assert len(temporary.execute()) == 0
```
Same, please avoid code duplication. This is the same as in the previous test, there are big blocks of code that are almost the same. Make generic code in a reusable function, and call it as needed, not this.

harsharaj96 replied:
Sure!

datapythonista commented on ibis/backends/tests/test_client.py:
```diff
@@ -202,3 +203,79 @@ def test_separate_database(con, alternate_current_database, current_data_db):
+
+    con.insert('temp_to_table', data_obj=records)
+    assert len(temporary.execute()) == 3
+    tm.assert_frame_equal(temporary.execute(), records)
```
In both cases we should add tests for at least the different values of `if_exists`. Otherwise we can't know if your implementation is working. Or if future changes break it.

harsharaj96 replied:
Done. Added separate tests for each of the values of `if_exists` for both the cases. Please take a look. Thanks.

datapythonista reviewed (commented):
Looks good in general, but we can't have code duplication as you're doing. See comments.

harsharaj96 commented:
Hi @datapythonista,
Did you get a chance to look at the changes?

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,120 @@ def _sqla_table_to_expr(self, table):
+            if if_exists == 'fail' and to_table_name in self.list_tables():
+                raise RuntimeError('The table already exists')
+            elif (
+                if_exists == 'replace' or if_exists == 'append'
```
`if_exists in ('replace', 'append')`

harsharaj96 replied:
Done!

datapythonista commented on ibis/backends/tests/test_client.py:
```diff
@@ -202,3 +203,263 @@ def test_separate_database(con, alternate_current_database, current_data_db):
     assert tmp_db.name == alternate_current_database
+
+
+def create_temp_table_with_schema(con, temp_table_name, schema):
```
`_create_temp_table_with_schema`  (starting with underscore) to better show that this is not a test, not something imported from anywhere.

harsharaj96 replied:
Done!

datapythonista commented on ibis/backends/tests/test_client.py:
```diff
@@ -202,3 +203,263 @@ def test_separate_database(con, alternate_current_database, current_data_db):
+        }
+    )
+
+    with pytest.raises(ValueError):
```
`raises` has a `match` argument that you can use to check that the error message is the one you expect. It's a regex, no need to compare the whole message if it's too long or doesn't make sense. But better to check that this is not failing for some other reason.

harsharaj96 replied:
Sure. One point to note is, `ValueError` is raised by newer versions of pandas (like https://pandas.pydata.org/pandas-docs/version/0.23.4/generated/pandas.DataFrame.to_sql.html) but not by older versions (like https://pandas.pydata.org/pandas-docs/version/0.18/generated/pandas.DataFrame.to_sql.html).

datapythonista commented on ibis/backends/tests/test_client.py:
```diff
@@ -202,3 +203,263 @@ def test_separate_database(con, alternate_current_database, current_data_db):
+            'first_name': ['A', 'B', 'C'],
+            'last_name': ['D', 'E', 'F'],
+            'department_name': ['AA', 'BB', 'CC'],
+            'salary': [100.0, 200.0, 300.0],
```
Would it makes sense to use always the same data, and have this, the loading on the data, and the assert that the right number of rows are there in the create table function you added?

I guess there is a reason I'm missing to have very similar data but different. But would make things much more compact and readable if we could have all the setup in the function, and in the tests just the specific things we want to test in each case.

harsharaj96 replied:
In some cases we want the `to_table` to be empty so that we can not only assert the number of rows in the final table but also the columns and other info using `assert_frame_equal`. `assert_frame_equal` can't be used when there are extra rows present compared to the `from_table`.  
I'll try to improvise the `_create_temp_table_with_schema` function to take in data as well but can't make the data part aslo go into this function since it needs to be reused.

datapythonista commented on ibis/backends/tests/test_client.py:
```diff
@@ -202,3 +203,263 @@ def test_separate_database(con, alternate_current_database, current_data_db):
+
+
+@pytest.mark.only_on_backends(
+    ['sqlite', 'postgres', 'mysql'],
```
Since we have now several tests, maybe we could define this in the top as a constant, like `SQLALCHEMY_BACKENDS = ['sqlite', 'postgres', 'mysql']`. Would make adding/removing backends much faster if needed in the future.

harsharaj96 replied:
Done!

datapythonista reviewed (commented):
Thanks for the updates to this PR @harsharaj96. Looks much better. Added few things that I think could still improve the code even more, but looks good now.

harsharaj96 commented:
Hi @datapythonista,
Added new changes as well. Have provided individual comments to a few of these changes please check them for more clarity.
Thanks!

datapythonista reviewed (approved):
Thanks @harsharaj96 very nice work.

@jreback, do you want to have a look?

datapythonista commented:
@harsharaj96 can you merge master into your branch please? Just realised the release notes are conflicting.

datapythonista commented:
Never mind, forgot I could quickly do it myself from the GitHub interface.

jreback commented:
will look

jreback commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,120 @@ def _sqla_table_to_expr(self, table):
+        self,
+        to_table_name: str,
+        database: Optional[str] = None,
+        data_obj=None,
```
can you type this?

is this signature consistent in naming? e.g. is `data_obj` what we call this generally?

harsharaj96 replied:
`data_obj` was earlier intended to send either pandas dataframe or a table expression. Later, we decided to only send the dataframe and if table expression needs to be used another variable with its workflow or some other change as suggested in this [issue](https://github.com/ibis-project/ibis/issues/2778) will be implemented.  
If the variable only uses dataframe we have used `data` as variable name in `.load_data()` function. So, I'll name it the same here.

jreback commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,120 @@ def _sqla_table_to_expr(self, table):
+            name of the table to which data needs to be inserted
+        database : string, optional
+            name of the attached database that the table is located in.
+        data_obj : pd.Dataframe
```
this is optional?

harsharaj96 replied:
Yes, but one of the optional parameters, either `data` or `from_table_name` must be sent by the user otherwise the function throws an error.

harsharaj96 replied:
Added `optional` to docstring and signature.

jreback commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,120 @@ def _sqla_table_to_expr(self, table):
+        data_obj : pd.Dataframe
+            data_obj is the dataframe containing data which needs to be
+            inserted to to_table_name
+        from_table_name : string, optional
```
umm, this is really strange to have 2 tables here?

harsharaj96 replied:
As mentioned in the previous comment, `from_table_name` is being used when `data` (dataframe) is not sent directly but when we want to insert data (all the data) from another table in the database.

datapythonista replied:
I thought you copied this signature from Impala, but I see you created your own.

I see that in Spark and Impala we've got the next signatures:
```
backends/spark/client.py:    def insert(
backends/spark/client.py-        self,
backends/spark/client.py-        table_name,
backends/spark/client.py-        obj=None,
backends/spark/client.py-        database=None,
backends/spark/client.py-        overwrite=False,
backends/spark/client.py-        values=None,
backends/spark/client.py-        validate=True,
backends/spark/client.py-    ):

backends/impala/client.py:    def insert(
backends/impala/client.py-        self,
backends/impala/client.py-        table_name,
backends/impala/client.py-        obj=None,
backends/impala/client.py-        database=None,
backends/impala/client.py-        overwrite=False,
backends/impala/client.py-        partition=None,
backends/impala/client.py-        values=None,
backends/impala/client.py-        validate=True,
backends/impala/client.py-    ):
```

Where did you get the signature you implemented from?

Happy to keep `if_exists` since you already implemented it. And I think we can ignore `values` for now, I think we should get rid of it, and pass values in `obj` instead. But for the rest, please be consistent with the signature of the other functions. Ideally users can write code for one backend, and unless there is a good reason, should work in all others.

harsharaj96 replied:
I was more or less trying to use the signature from impala itself.  
After a couple of requested changes, I did drop the `values` parameter and its usage. I've also updated the `data_obj` parameter to `obj` like in impala.  
I couldn't use `table_name` as is because we added a separate workflow if a table was passed (`from_table_name`) for inserting data into a table (`to_table_name`). If it's fine, shall I update the `to_table_name` parameter to `table_name`?  
Let me check how `overwrite` is being used in impala and update `if_exists` to it. Since `overwrite` is a boolean and `if_exists` is not (it's a string) do you want me to make `overwrite` a string parameter or update the code to use it as a boolean?

harsharaj96 replied:
Updated `to_table_name` variable to `table_name`.

harsharaj96 replied:
`if_exists` is also being used in the `load_data` function of the SQLAlchemy backend but not `overwrite`, though `overwrite` is being used in Impala. So, that was the reason behind my implementation having `if_exists` and not `overwrite`.  
It's not a big task to modify the functionality and test cases. Please let me know what we want to go forward with @datapythonista.
Thanks!

datapythonista replied:
So, the idea would be that if an Ibis user wrote a call to `insert`  in Impala, and changes backend to postgres, the code should still work. This is not always 100% possible, for example, the `partition` parameter is specific to Impala. But for the rest, would be good to keep all `insert` methods consistent.

So, would be good to change:
- Merge `data` and `from_table_name` into a single `obj` parameter, and then check the type to see which operation has to be done.
- Replace `if_exists` by `overwrite` as a boolean. We will lose the `fail` case, but the users can easily check themselves `if con.table('whatever').exists():` or similar to implement that.
- Change the order of parameters so it matches the order of the Impala/Spark `insert`
- Any other change that makes sense

I think it should be quite quick to change this now that you've got your implementation working. Thanks for all the hard work in this PR, and apologies for not realizing on this earlier, and incorrectly assumed that you were using the other `insert` methods as a reference.

harsharaj96 replied:
Thanks for the inputs @datapythonista.  
It was my mistake as well to not implement the functionality similar to impala ones and check with you.  
Anyways, I've updated the `insert` function in SQLAlchemy, and the test functions with the above suggestions. Please take a look and let me know if there's anything I'm still missing.
Thanks once again!

jreback commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,120 @@ def _sqla_table_to_expr(self, table):
+            and the from_table_name (table). Please use only one
+            parameter.
+
+        RuntimeError
```
why not a ValueError? is this typically for other ops in sql alchemy?

harsharaj96 replied:
Agreed. Even pandas raises `ValueError` when `if_exists` is given as 'fail' while using `to_sql` function. Will update it.

jreback commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,120 @@ def _sqla_table_to_expr(self, table):
+            from_table_name, str
+        ):
+            raise ValueError(
+                "Sorry, can't insert from both the data_obj (dataframe)"
```
use a more declarative language. e.g.

`Cannot insert from both....` 

also in-line the paramters via f-strings from what was passed (if you its a table name for example)

harsharaj96 replied:
Done!

jreback commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +329,120 @@ def _sqla_table_to_expr(self, table):
+                            )
+                        )
+        else:
+            raise TypeError(
```
why is this a TypeError?

harsharaj96 replied:
Because we are expecting `from_table_name` to be of type `str` and if anything else was given we wanted to raise an error (based on discussion during previous reviews).

jreback reviewed (changes requested):
i find the signature a bit odd

harsharaj96 commented:
Hi @jreback,
Added new changes as well. Have provided individual comments to a few of these changes please check them for more clarity.
Thanks!

harsharaj96 commented:
Hi @datapythonista/@jreback,
Can you please check and approve the changes so that checks can be performed? I've added a couple of changes based on the last failing checks. Just wanted to see if they are working fine.
Thanks!

datapythonista commented:
@harsharaj96 in case you haven't seen it, tests run and failed.

harsharaj96 commented:
@datapythonista ,
So sorry, might have missed it. Added a main bug fix. Please approve for a re-run of tests.

harsharaj96 commented:
Hi @datapythonista,
Added a couple of more bug fixes, please check and make the workflow run the checks and tests.
Thanks!

harsharaj96 commented:
Hi @datapythonista & @jreback,
All the checks and tests have passed. Please take a look and approve the pull request if you think everything is fine. Thanks a lot for your inputs and suggestions.
Cheers!

harsharaj96 commented:
Hi @datapythonista & @jreback,  
Please go through the changes and let me know if I can help you with anything else.  
Thanks!

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +330,117 @@ def _sqla_table_to_expr(self, table):
+                    table_name,
+                    self.con,
+                    index=False,
+                    if_exists='replace',
```
You don't need to repeat everything, you can remove the if and use:

```suggestion
                    if_exists='replace' if orverwrite else 'append',
```

harsharaj96 replied:
Done!

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +330,117 @@ def _sqla_table_to_expr(self, table):
+                    self.con,
+                    index=False,
+                    if_exists='replace',
+                    **params,
```
Would it work if we use `schema=self.database_name` here, when there is no attachment?

harsharaj96 replied:
If I remember correctly, like mentioned in another conversation, tests were erroring out when the database was not set to `None`. So, I'm not sure if it works. I was able to resolve things by following the code in the `.load_data()` function where the same conditionality and setting for `params` were used.

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +330,117 @@ def _sqla_table_to_expr(self, table):
+            raise TypeError(
+                'No operation is being performed. Either the obj'
+                ' parameter is not a pandas dataframe or'
+                ' is not of string datatype.'
```
Or an Ibis `TableExr` now, right?

harsharaj96 replied:
Yes. My bad missed it in the message. Updated!

datapythonista commented on ibis/backends/tests/test_client.py:
```diff
@@ -202,3 +205,245 @@ def test_separate_database(con, alternate_current_database, current_data_db):
+            'department_name': ['XX', 'YY', 'ZZ'],
+            'salary': [400.0, 500.0, 600.0],
+        }
+    )
```
There is a lot of repetition in these tests. Like this dataframe or similar being repeated lots of times. Can you define a fixture or a constant, and try to use always the same data? If you need different data (but similar) you can make a copy of the dataframe, and change whatever you need).

Same for the schema. And anything that keeps repeating among tests. Write reusable code, and don't copy-paste anything, which is a very poor programming practice.

harsharaj96 replied:
Sure, will work on it!

datapythonista reviewed (commented):
Thanks a lot for the updates, and your flexibility, and all the hard work in this PR. Looks good, still added few more comments.

harsharaj96 commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +330,117 @@ def _sqla_table_to_expr(self, table):
+
+        if database == self.database_name:
+            # avoid fully qualified name
+            database = None
```
I did it based on the other functions like `.create_table()`, `.load_data()`, `.drop_table()` in SQLAlchemy. So I thought this to be a requirement. Also, if I remember correctly, keeping the database name intact and not `None` was giving errors in the tests.

harsharaj96 commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +330,117 @@ def _sqla_table_to_expr(self, table):
+
+        if database is not None:
+            raise NotImplementedError(
+                'Inserting data to a table from a different database is not '
```
I don't think so. Even `.create_table()`, `.load_data()`, `.drop_table()` in SQLAlchemy backend have these conditions.

harsharaj96 commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +330,117 @@ def _sqla_table_to_expr(self, table):
+                'yet implemented'
+            )
+
+        if obj is None:
```
Agree. Updated!

harsharaj96 commented:
Hi @datapythonista,
I've added new changes based on your suggestions. Added a few replies to some of the comments and implemented a few. Please take a look and let me know.
Thanks!

datapythonista commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +330,101 @@ def _sqla_table_to_expr(self, table):
+            raise ValueError(
+                f"No operation is being performed. Either the obj parameter"
+                f" is not a pandas DataFrame or is not of string data type for"
+                f"sending table name or is not a ibis Table Expression."
```
No big deal, but the `f` is only needed in the lines where there is a variable, not all them.

harsharaj96 replied:
Fixed!

datapythonista reviewed (commented):
Thanks for the updates, and really nice clean up of the tests, this is way better.

Looks good, @jreback do you want to have a look?

jreback commented:
i will look soon

jreback commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +330,101 @@ def _sqla_table_to_expr(self, table):
+    def insert(
+        self,
+        table_name: str,
+        obj=None,
```
how can this ever be None? pls fully type this. Why allow a string at all? that is not supported anywhere else in the ibis api.

jreback replied:
let's start really simple and just take a DataFrame or an Expr

harsharaj96 replied:
Done!

jreback commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +330,101 @@ def _sqla_table_to_expr(self, table):
+            )
+
+        params = {}
+        if self.has_attachment:
```
what is this?

harsharaj96 replied:
This is used to Attach another database file if there is any. Followed it from the implementation in `.load_data` function

jreback commented on ibis/backends/tests/conftest.py:
```diff
@@ -306,3 +307,45 @@ def alternate_current_database(con, backend, current_data_db: str) -> str:
         con.drop_database(name, force=True)
+
+
+@pytest.fixture
```
are these *only* used in the new tests? if so i would just put them in the sql tests themselves for now. alt these can stay but give real names to these

harsharaj96 replied:
I was following the naming convention given for the other fixtures. Sure, I'll update the names.

harsharaj96 replied:
Done!

jreback commented on ibis/backends/tests/test_client.py:
```diff
@@ -202,3 +205,128 @@ def test_separate_database(con, alternate_current_database, current_data_db):
+    return temporary
+
+
+@pytest.mark.only_on_backends(
```
these are not fully generic tests as the .insert api is not generic. move to the sql alchemy test suite

harsharaj96 replied:
@datapythonista wanted the tests to be present here so that we can later use these tests for all the backends once the implementations are complete.

jreback reviewed (changes requested):
i would prefer to actually adjust the signature for all backends to make this generic.

harsharaj96 commented:
Hi @jreback & @datapythonista,
I've added new changes based on your suggestions. Added a few replies to some of the comments and implemented a few. Please take a look and let me know.
Thanks!

jreback commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -329,3 +330,93 @@ def _sqla_table_to_expr(self, table):
+    def insert(
+        self,
+        table_name: str,
+        obj,
```
can you type obj

harsharaj96 replied:
Done!

jreback reviewed (changes requested):
small comment, pls merge master and ping on greenish (the 3.7 tests are still failing) but that's ok

harsharaj96 commented:
Hi @jreback,

jreback commented:
thanks @harsharaj96

datapythonista commented:
Thanks a lot for all your work on this @harsharaj96, great job!
